### PR TITLE
Newnnet3

### PR DESCRIPTION
**Pull request by mrsrikanth**
Hi @danpovey ,

Here is the PR for the chain2 recipes.

Srikanth

### CODE REVIEW

**Comment by francisr** [comment]
What are the differences with the current chain training?

**Comment by danpovey** [comment]
Script-level differences in how egs are dumped and dealt with, intended to
make it easier to do things like incremental or multilingual training.

File: egs/swbd/s5c/RESULTS [hunk 1]
```diff
@@ -247,3 +247,7 @@ exit 0
 %WER 14.6 | 1831 21395 | 87.1 9.2 3.6 1.8 14.6 51.8 | exp/tri7uc-sat_mmi_b0.1/decode_eval2000_graph_sw1_tg_it4/score_17_0.0/eval2000.ctm.swbd.filt.sys
 # fisher 4gram rescoring,
 %WER 13.2 | 1831 21395 | 88.3 8.2 3.4 1.5 13.2 49.2 | exp/tri7uc-sat_mmi_b0.1/decode_eval2000_graph_sw1_fsh_fg_it4/score_19_0.0/eval2000.ctm.swbd.filt.sys
+
+# New nnet3 recipe with tdnn_7k
```
Comment (jtrmal) [inline egs/swbd/s5c/RESULTS]: in all the results, would it be possible to keep the order with tg/fg and other modifiers as the previous results, please?

Comment (mrsrikanth) [reply egs/swbd/s5c/RESULTS]: Hi @jtrmal, the result file has been updated.

**Comment by rezame** [comment]
Hi
can change "new" to something that have more meaning and concept.
In feature the new goes old! and we have newer and more newer and ... models

**Comment by mrsrikanth** [comment]
@rezame sure, I can change all "new" to "chain2". But I believe the mentions are mostly in comments and RESULTS files.

File: egs/wsj/s5/steps/nnet3/chain2/process_egs.sh [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+  exit 1;
+fi
+
+raw_egs_dir=$1
+dir=$2
+
+# die on error or undefined variable.
+set -e -u
+
+if ! steps/chain/validate_raw_egs.sh $raw_egs_dir; then
```
Comment (desh2608) [inline egs/wsj/s5/steps/nnet3/chain2/process_egs.sh]: Should this be `steps/nnet3/chain2/validate_raw_egs.sh`?

Comment (mrsrikanth) [reply egs/wsj/s5/steps/nnet3/chain2/process_egs.sh]: yes! I have fixed this and will be pushing the changes with the BABEL recipe soon. I should probably run the tests for mini_librespeech again.

File: egs/wsj/s5/steps/nnet3/chain2/process_egs.sh [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+# This script takes nnet examples dumped by steps/chain/get_raw_egs.sh and
+# combines the chunks into groups by speaker (to the extent possible; it may
+# need to combine speakers in some cases), locally randomizes the result, and
+# dumps the resulting egs to disk.  Chunks of these will later be globally
+# randomized (at the scp level) by steps/chaina/randomize_egs.sh
+
+
+# Begin configuration section.
+cmd=run.pl
+chunks_per_group=4
```
Comment (danpovey) [inline egs/wsj/s5/steps/nnet3/chain2/process_egs.sh]: Hm.  I think for normal models this chunks_per_group should default to 1, and num_repeats should default to 1 too.  Grouping utterances by speaker was necessary for some models I was investigating at the time, but not now.

File: egs/wsj/s5/steps/nnet3/chain2/process_egs.sh [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+echo "$0 $@"  # Print the command line for logging
+
+if [ -f path.sh ]; then . ./path.sh; fi
+. parse_options.sh || exit 1;
+
+
+if [ $# != 2 ]; then
+  echo "Usage: $0 [opts] <raw-egs-dir> <processed-egs-dir>"
+  echo " e.g.: $0 --chunks-per-group 4 exp/chaina/tdnn1a_sp/raw_egs exp/chaina/tdnn1a_sp/processed_egs"
```
Comment (danpovey) [inline egs/wsj/s5/steps/nnet3/chain2/process_egs.sh]: If I understand correctly, the chunks-per-group actually does nothing,and neither does num-repeats.  Can you make sure they are completely removed, including from the diagnostic part below?  (There is an awk script).

File: src/nnet3/nnet-chain-training2.h [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
+// WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef KALDI_NNET3_NNET_CHAIN_TRAINING_H_
```
Comment (danpovey) [inline src/nnet3/nnet-chain-training2.h]: please fix include guard.  There is a script in misc/ that is for that, you might find other bad ones.

File: egs/wsj/s5/steps/nnet3/chain2/process_egs.sh (hunk 3 above)
Comment (mrsrikanth) [reply egs/wsj/s5/steps/nnet3/chain2/process_egs.sh]: I have fixed this in the latest commit.

File: egs/wsj/s5/steps/nnet3/chain2/process_egs.sh (hunk 2 above)
Comment (mrsrikanth) [reply egs/wsj/s5/steps/nnet3/chain2/process_egs.sh]: this is fixed in the latest commit.

File: src/nnet3/nnet-chain-training2.h (hunk 5 above)
Comment (mrsrikanth) [reply src/nnet3/nnet-chain-training2.h]: I modified it manually. When I ran misc/maintenance/fix_include_guards.sh, it changed a whole bunch of files not related to this PR.

**Comment by danpovey** [comment]
@mrsrikanth how long do you think it will be before you have a PR with your multilingual recipes?

File: egs/babel_multilang/s5/local/chain2/run_tdnn.sh [hunk 6]
```diff
@@ -405,27 +408,23 @@ fi
...
+           nnet3-init --srand=${srand} ${dir}/configs/final.config ${dir}/init/multi.raw || exit 1
 fi
 
 if [ $stage -le 18 ]; then
   echo "$0: Starting model training"
   steps/chain2/train.sh \
     --stage $train_stage --cmd "$cuda_cmd" \
     --multilingual-eg true \
-    --xent-regularize $xent_regularize --leaky-hmm-coefficient 0.1 \
+    --xent-regularize $xent_regularize --leaky-hmm-coefficient 0.25 --out-of-range-regularize 0.0 \
```
Comment (danpovey) [inline egs/babel_multilang/s5/local/chain2/run_tdnn.sh]: Was the out-of-range-regularize 0.0 necessary?  Or a debugging thing?

Comment (mrsrikanth) [reply egs/babel_multilang/s5/local/chain2/run_tdnn.sh]: No, I don't think it is necessary to set it to 0.0. It was for debugging purposes only. I can remove it and test it again.

File: egs/babel_multilang/s5/local/chain2/run_tdnn.sh [hunk 7]
```diff
@@ -37,8 +37,9 @@ tree_affix=  # affix for tree directory, e.g. "a" or "b", in case we change the
 tdnn_affix=  #affix for TDNN directory, e.g. "a" or "b", in case we change the configuration.
 feat_suffix=_hires      
 
+label_delay=5
 frame_subsampling_factor=3
-xent_regularize=0.1
+xent_regularize=0.025
```
[the remark this answers is not in the document]
Comment (mrsrikanth) [reply egs/babel_multilang/s5/local/chain2/run_tdnn.sh]: No, it is not necessary. I can switch it back to 0.1.

[the remark this answers is not in the document]
Comment (mrsrikanth) [reply egs/babel_multilang/s5/local/chain2/run_tdnn.sh]: I made the change.

File: egs/babel_multilang/s5/local/chain2/run_tdnn.sh (hunk 6 above)
Comment (mrsrikanth) [reply egs/babel_multilang/s5/local/chain2/run_tdnn.sh]: I removed it and change is pushed.

**Comment by danpovey** [comment]
@mrsrikanth please not the conflict!

File: egs/babel_multilang/s5/local/chain2/run_tdnn.sh (hunk 6 above)
Comment (mrsrikanth) [reply egs/babel_multilang/s5/local/chain2/run_tdnn.sh]: > @mrsrikanth please not the conflict!

@danpovey I don't understand, what conflict are you referring to?

**Comment by danpovey** [comment]
The git conflict.  in a Makefile.

**Comment by danpovey** [comment]
Merging via #3981

**Pull request closed without merging**